pdpserver/server/ctrl_storage.go

```diff
@@ -101,7 +101,9 @@ func (handler *storageHandler) ServeHTTP(w http.ResponseWriter, r *http.Request)
 
 	switch cmd {
 	case queryCmd:
+		handler.s.RLock()
```

Comment:
That's too much to lock. With the lock you can't update policy during whole request handling. You need to lock only storage pointer read (just one quick operation):
```go
handler.s.RLock()
p := handler.s.p
handler.s.RUnlock()
handleQuery(w, p, resourcePath, urlQuery)
```
Read access for storage itself doesn't require locking as it protected by copy-on-write mechanism. What is important to protect is read the pointer to storage.